### Prepare 

In [1]:
import pandas as pd
%pylab inline
from datetime import datetime
from datetime import timedelta

# date handling
dateparse1 = lambda x: pd.datetime.strptime(x, '%d.%m.%Y')
dateparse2 = lambda x: pd.datetime.strptime(x, '%m/%d/%Y')

Populating the interactive namespace from numpy and matplotlib


## Load Data

In [2]:
piDat = pd.read_csv('../data/Webtrekk_PI-V.csv', sep=';', parse_dates = ['VB Tag','VE Tag'], date_parser = dateparse1)
abo_canc = pd.read_csv("../data/SSOID_CancDatum.CSV", encoding = 'latin-1', 
                       date_parser = dateparse2, parse_dates=["K Tag"], dtype={'AN Promotion ID':np.int32})
# remove spaces from column
abo_canc['VT Kündigungsgrund'] = abo_canc['VT Kündigungsgrund'].apply(lambda x: x.strip())

In [3]:
sorted(abo_canc["K Tag"], reverse = True)


[Timestamp('2018-04-10 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-09 00:00:00'),
 Timestamp('2018-04-

In [4]:
print(piDat.shape)
print(piDat['WT SSOID'].unique().shape)

(99327, 27)
(13714,)


<span style="color:blue"> Doppelte Einträge für SSOID und Jahr/KW: </span>

In [5]:
print(piDat.loc[piDat[['WT SSOID','WT Jahr/KW']].duplicated(keep = False)].sort_values('WT SSOID').shape)
piDat.loc[piDat[['WT SSOID','WT Jahr/KW']].duplicated(keep = False)].sort_values('WT SSOID')

(1766, 27)


WT SSOID WT Jahr/KW  WT Jahr  WT KW  WT Visits  WT Visits Plus  \
267    ASAG1693934     2018/4     2018      4         37              17   
268    ASAG1693934     2018/4     2018      4         37              17   
852    ASAG1772044    2017/42     2017     42          2               0   
853    ASAG1772044    2017/42     2017     42          2               0   
854    ASAG1772044    2017/42     2017     42          2               0   
2227   ASAG1960328    2017/38     2017     38          5               2   
2228   ASAG1960328    2017/38     2017     38          5               2   
2252   ASAG1960653    2017/39     2017     39         38               7   
2251   ASAG1960653    2017/39     2017     39         38               7   
2249   ASAG1960653    2017/37     2017     37         42              16   
2248   ASAG1960653    2017/37     2017     37         42              16   
2240   ASAG1960653    2017/36     2017     36         55               9   
2239   ASAG1960653    2017/36     2017     36         55               9   
2236   ASAG1960653    2017/38     2017     38         58              13   
2235   ASAG1960653    2017/38     2017     38         58              13   
2673   ASAG1969906    2017/45     2017     45         34              13   
2674   ASAG1969906    2017/50     2017     50         21              11   
2675   ASAG1969906    2017/50     2017     50         21              11   
2676   ASAG1969906    2017/43     2017     43         21               7   
2677   ASAG1969906    2017/43     2017     43         21               7   
2678   ASAG1969906    2017/49     2017     49         31              18   
2679   ASAG1969906    2017/49     2017     49         31              18   
2687   ASAG1969906    2017/44     2017     44         15               6   
2681   ASAG1969906    2017/51     2017     51         29              12   
2682   ASAG1969906    2017/42     2017     42         28               4   
2683   ASAG1969906    2017/42     2017     42         28               4   
2684   ASAG1969906    2017/52     2017     52         28               7   
2685   ASAG1969906    2017/52     2017     52         28               7   
2686   ASAG1969906    2017/44     2017     44         15               6   
2672   ASAG1969906    2017/45     2017     45         34              13   
...            ...        ...      ...    ...        ...             ...   
98919  BILD1880752    2017/48     2017     48          3               3   
98920  BILD1880752    2017/39     2017     39          2               2   
98918  BILD1880752    2017/48     2017     48          3               3   
98922  BILD1880752    2017/42     2017     42          1               1   
98923  BILD1880752    2017/42     2017     42          1               1   
98924  BILD1880752    2017/37     2017     37          5               4   
98925  BILD1880752    2017/37     2017     37          5               4   
98921  BILD1880752    2017/39     2017     39          2               2   
98964  BILD1903484    2017/43     2017     43          1               0   
98963  BILD1903484    2017/43     2017     43          1               0   
98962  BILD1903484    2017/43     2017     43          1               0   
98961  BILD1903484    2017/40     2017     40          1               0   
98960  BILD1903484    2017/40     2017     40          1               0   
98959  BILD1903484     2018/1     2018      1          2               0   
98957  BILD1903484    2017/48     2017     48          2               0   
98956  BILD1903484    2017/48     2017     48          2               0   
98955  BILD1903484    2017/48     2017     48          2               0   
98954  BILD1903484     2018/5     2018      5          2               0   
98953  BILD1903484     2018/5     2018      5          2               0   
98958  BILD1903484     2018/1     2018      1          2               0   
99014  BILD2001976    2017/38     2017     38         22  

<span style = "color:blue"> werden einfach alle ignoriert </span>

In [3]:
pi_wo_dup = piDat.drop_duplicates(piDat[['WT SSOID','WT Jahr/KW']],keep = False)

In [4]:
# make single tables for visit, pi, etc. (col_names=KW/Jahr, index = SSOID)
pi_per_ssoid = pi_wo_dup.pivot(index='WT SSOID', columns='WT Jahr/KW', values='WT Page Impressions')
v_per_ssoid = pi_wo_dup.pivot(index='WT SSOID', columns='WT Jahr/KW', values='WT Visits')
piPlus_per_ssoid = pi_wo_dup.pivot(index='WT SSOID', columns='WT Jahr/KW', values='WT Page Impressions plus')
vPlus_per_ssoid = pi_wo_dup.pivot(index='WT SSOID', columns='WT Jahr/KW', values='WT Visits Plus')

In [8]:
# Verhältnisse bilden
pi_per_v = pi_per_ssoid / v_per_ssoid
piPlus_per_v = piPlus_per_ssoid / v_per_ssoid
vPlus_per_v = vPlus_per_ssoid / v_per_ssoid

## Vertragsdaten filtern und vorbereiten

In [5]:
# contract details only
contract = pi_wo_dup[['WT SSOID', 'Kuendigungsstatus', 'Kuendigungsgrund ID',
       'Kuendigungsgrund', 'Angebotsgruppe', 'Promotion ID',
       'Angebotsname', 'Vertragsform', 'Trial Kennzeichen',
       'Vertragsphase', 'VB Tag', 'VB Jahr', 'VB KW', 'VB Jahr/KW',
       'VE Tag', 'VE Jahr', 'VE KW', 'VE Jahr/KW']].drop_duplicates()

In [10]:
print(pi_wo_dup.shape)
print(contract.shape)

(97561, 27)
(14214, 18)


In [11]:
# filter: Kuendigungsgrund
contract.groupby('Kuendigungsgrund').size()

Kuendigungsgrund
Doppelkauf                                    127
Falsche Erfassung/Auftragsanlage                3
Finanzielle Gründe                            548
Gefällt nicht - redaktionell Artikelmenge     980
Gefällt nicht - redaktionell Qualität          85
Gefällt nicht- technisch                       40
KAW mit neuem Produkt / man. Downgrade         13
Keine Zeit                                     23
Kreditkartengültigkeit abgelaufen              87
Kunde kündigt, da er aktuell nicht nutzt       49
Kunde verstorben                                4
Kündigung mit Vollerstattung                    6
Nichtzahler                                   996
Ohne Angabe eines Grundes                    4343
Technische Fehler                               9
Testbestellung                                 16
Widerruf                                      438
Zu teuer                                       18
Name: WT SSOID, dtype: int64

In [6]:
# folgende Kuendingunsgrunde werden rausgefiltert
KuendigungNotUsed = ['Doppelkauf', 'Falsche Erfassung/Auftragsanlage', 'Kreditkartengültigkeit abgelaufen',
                    'Kunde verstorben', 'Nichtzahler', 'Technische Fehler', 'Testbestellung', 'Widerruf',
                    'KAW mit neuem Produkt / man. Downgrade', 'Kündigung mit Vollerstattung']
contract_Kuendigung = contract.loc[-contract['Kuendigungsgrund'].isin(KuendigungNotUsed)]

In [7]:
# Ende vom letzten Vertrag mit VollAbo
last_contract_end = contract_Kuendigung.loc[contract_Kuendigung['Vertragsphase']=='Vollabo'].sort_values('VE Jahr/KW',ascending=False).groupby('WT SSOID')['VE Jahr/KW'].first()
second_last_contract_end = contract_Kuendigung.loc[contract_Kuendigung['Vertragsphase']=='Vollabo'].sort_values('VE Jahr/KW',ascending=False).groupby('WT SSOID')['VE Jahr/KW'].nth(2)
third_last_contract_end = contract_Kuendigung.loc[contract_Kuendigung['Vertragsphase']=='Vollabo'].sort_values('VE Jahr/KW',ascending=False).groupby('WT SSOID')['VE Jahr/KW'].nth(3)

In [14]:
print(last_contract_end.shape)
print(second_last_contract_end.shape)
print(third_last_contract_end.shape)

(8658,)
(1,)
(0,)


Es gibt 8658 Verträge, die unsere Bedingungen erfüllen: 
- Vollabo
- kein komischer Kündigungsgrund

Ein Nutzer hat zwei von diesen Verträgen. Wird einfach ignoriert.

In [8]:
# merge with all contract data
last_contract_end=last_contract_end.to_frame().reset_index()
last_contract = pd.merge(last_contract_end, contract_Kuendigung[['WT SSOID', 'Kuendigungsstatus','Kuendigungsgrund','Promotion ID','Vertragsform', 'Trial Kennzeichen',
       'Vertragsphase', 'VB Tag', 'VB Jahr', 'VB KW', 'VB Jahr/KW',
       'VE Tag', 'VE Jahr', 'VE KW', 'VE Jahr/KW']], on=['WT SSOID', 'VE Jahr/KW'])

In [9]:
# merge with abo_canc to get K Tag
last_contract_KTag = pd.merge(last_contract,abo_canc[['KD SSO-ID','AN Promotion ID', 'K Tag',
       'VT Kündigungsgrund']], left_on=['WT SSOID','Promotion ID','Kuendigungsgrund'], 
         right_on=['KD SSO-ID','AN Promotion ID','VT Kündigungsgrund'],how='left')

In [10]:
print(last_contract.shape)
print(last_contract_KTag.shape)

(8658, 15)
(8845, 19)


In [11]:
# doppelte Einträge in last_contract_KTag nach merge mit abo_canc werden ignoriert
last_contract_KTag = last_contract_KTag.drop_duplicates(last_contract_KTag[['WT SSOID','Promotion ID','Kuendigungsgrund',
                                        'KD SSO-ID','AN Promotion ID','VT Kündigungsgrund']], keep=False)

In [19]:
idx = last_contract_KTag['WT SSOID']

# merge with pi, v, pi_perv etc. to only get SSOIDs of interest in these tables
pi_per_ssoid_cont = pi_per_ssoid.loc[pi_per_ssoid.index.isin(idx)]
v_per_ssoid_cont = v_per_ssoid.loc[v_per_ssoid.index.isin(idx)]
piPlus_per_ssoid_cont = piPlus_per_ssoid.loc[piPlus_per_ssoid.index.isin(idx)]
vPlus_per_ssoid_cont = vPlus_per_ssoid.loc[vPlus_per_ssoid.index.isin(idx)]

pi_per_v_cont = pi_per_v.loc[pi_per_v.index.isin(idx)]
piPlus_per_v_cont = piPlus_per_v.loc[piPlus_per_v.index.isin(idx)]
vPlus_per_v_cont = vPlus_per_v[vPlus_per_v.index.isin(idx)]

## Ein paar Ungereimtheiten

Gekuendigt, aber kein Kuendigungsdatum? (VE Tag 4.5.18) UNGEWÖHNLICH. Nochmal nachgucken

In [20]:
last_contract_KTag.loc[last_contract_KTag['WT SSOID']=='ASAG5295080']
last_contract_KTag.loc[last_contract_KTag['WT SSOID']=='ASAG5295080']

WT SSOID VE Jahr/KW Kuendigungsstatus           Kuendigungsgrund  \
1403  ASAG5295080    2018/14        gekuendigt  Ohne Angabe eines Grundes   

      Promotion ID Vertragsform  Trial Kennzeichen Vertragsphase     VB Tag  \
1403      10003002          NOP  kostenloser Trial       Vollabo 2017-12-06   

      VB Jahr  VB KW VB Jahr/KW     VE Tag  VE Jahr  VE KW KD SSO-ID  \
1403     2017     50    2017/49 2018-04-05     2018     14       NaN   

      AN Promotion ID K Tag VT Kündigungsgrund  
1403              NaN   NaT                NaN

In [21]:
sum((last_contract_KTag['Kuendigungsstatus'] == 'gekuendigt') & (pd.isnull(last_contract_KTag['K Tag']))) 

2

## Metriken zum Angucken der Verteilungen

In [22]:
last_contract_KTag.loc[last_contract_KTag['WT SSOID']=='ASAG5295080']

WT SSOID VE Jahr/KW Kuendigungsstatus           Kuendigungsgrund  \
1403  ASAG5295080    2018/14        gekuendigt  Ohne Angabe eines Grundes   

      Promotion ID Vertragsform  Trial Kennzeichen Vertragsphase     VB Tag  \
1403      10003002          NOP  kostenloser Trial       Vollabo 2017-12-06   

      VB Jahr  VB KW VB Jahr/KW     VE Tag  VE Jahr  VE KW KD SSO-ID  \
1403     2017     50    2017/49 2018-04-05     2018     14       NaN   

      AN Promotion ID K Tag VT Kündigungsgrund  
1403              NaN   NaT                NaN

In [32]:
# calculating metrics for inspecting user behaviour
def calc_metrics(dat):
    
   # print(dat.name)
    
    # delete user behavior after cancellation in order to get clean data
    if (last_contract_KTag.loc[last_contract_KTag['WT SSOID']==dat.name,'Kuendigungsstatus'] == 'gekuendigt').bool():
        # cancellation day is K Tag vrom contract data
        canc_day = last_contract_KTag.loc[last_contract_KTag['WT SSOID']==dat.name,'K Tag']
        # if cancellation day is not set, set VE Tag
        if pd.isnull(canc_day).bool():
            canc_day = last_contract_KTag.loc[last_contract_KTag['WT SSOID']==dat.name,'VE Tag']
            
        canc_day = canc_day + timedelta(days=7)
        
        # if cancellation date is before webtrekk-data, delete all entries
        if (canc_day < pd.datetime(2017,9,4)).bool():
            for idx, val in enumerate(dat):
                dat.iloc[idx] = np.nan
        # otherwise delete activity after cancellation day    
        else:
            p = pd.PeriodIndex(freq='w', start=canc_day.dt.date.item(), end=pd.datetime(2018,2,28))
            for idx,val in enumerate(p):
                idxStr = str(val.year) + '/' + str(val.weekofyear)
                dat[idxStr] = np.nan
    
    
    ### STANDARD METRICS
    
    # mean
    mean = dat.mean() 
    # median
    median = dat.median()
    # sd
    sd = dat.std()
    # number of active weeks
    active = len(dat) - sum(dat.isna())
    # min
    mini = dat.min()
    # max
    maxi = dat.max()
    
    # write data
    dat['mean'] = mean
    dat['median'] = median
    dat['sd'] = sd
    dat['activ_wks'] = active
    dat['min'] = mini
    dat['max'] = maxi
    dat['diff_max_min'] = maxi - mini

    
    
    ### USER BEHAVIOR BEFORE CANCELLATION
    
    # find user behaviour before cancellation, if there is no cancellation return last months  
    # if there is no cancellation just use user behaviour of last three months
    if (last_contract_KTag.loc[last_contract_KTag['WT SSOID']==dat.name,'Kuendigungsstatus'] == 'ungekuendigt').bool() :
        
        if sum(~pd.isnull(dat)) < 4:
            
            # user behaviour last 4 weeks
            dat['usr_behav4'] = np.nan
            dat['usr_behav3'] = np.nan
            dat['usr_behav2'] = np.nan
            dat['usr_behav1'] = np.nan
            
        else:    
            
            # get active weeks = non-zero values 
            actWks = np.nonzero(~pd.isnull(dat.values))
            # sample randomly values from active weeks
            rdmSmp = np.random.choice(actWks[0],4)
            
            # user behaviour last 4 weeks
            dat['usr_behav4'] = rdmSmp[0]
            dat['usr_behav3'] = rdmSmp[1]
            dat['usr_behav2'] = rdmSmp[2]
            dat['usr_behav1'] = rdmSmp[3]

    else:
        
        canc_day = last_contract_KTag.loc[last_contract_KTag['WT SSOID']==dat.name,'K Tag']
        
        # cancellaton date is too early, there is not enough information on user behaviour
        # --> setting values to Na
        canc_jahr = canc_day.dt.year.item()
        canc_kw =canc_day.dt.week.item()

        if (canc_jahr <= 2017) & (canc_kw < 39):

            # user behaviour last 4 weeks
            dat['usr_behav4'] = np.nan
            dat['usr_behav3'] = np.nan
            dat['usr_behav2'] = np.nan
            dat['usr_behav1'] = np.nan
      
        
        else:
                    
            # if cancellation day is not set
            if pd.isnull(canc_day).bool():
                canc_day = last_contract_KTag.loc[last_contract_KTag['WT SSOID']==dat.name,'VE Tag']
                canc_jahr = canc_day.dt.year.item()
                canc_kw =canc_day.dt.week.item()
                
            # if cancellation is too recent, substract one week from canc_day
            if (canc_jahr == 2018) & (canc_kw > 9):
                weeks = canc_kw - 9
                canc_day =  canc_day -  pd.to_timedelta(weeks*7, unit='d')
            
            # date series of last 4 weeks before cancellation date
            p = pd.PeriodIndex(freq='w', end=canc_day.dt.date.item(), periods=4)
            
            # construct correct column index from date series
            # user behavior last 4 weeks
            dat['usr_behav4'] = dat[ str(p[3].year)+'/'+str(p[3].weekofyear) ]
            dat['usr_behav3'] = dat[ str(p[2].year)+'/'+str(p[2].weekofyear) ]
            dat['usr_behav2'] = dat[ str(p[1].year)+'/'+str(p[1].weekofyear) ]
            dat['usr_behav1'] = dat[ str(p[0].year)+'/'+str(p[0].weekofyear) ]
            
            
    # metrics to measure change of user behavior before cancellation
    dat['prz_usr_behav4'] = dat['usr_behav4']/median
    dat['prz_usr_behav3'] = dat['usr_behav3']/median
    dat['prz_usr_behav2'] = dat['usr_behav2']/median
    dat['prz_usr_behav1'] = dat['usr_behav1']/median
    
    dat['zsc_usr_behav4'] = (dat['usr_behav4']-mean)/sd
    dat['zsc_usr_behav3'] = (dat['usr_behav3']-mean)/sd
    dat['zsc_usr_behav2'] = (dat['usr_behav2']-mean)/sd
    dat['zsc_usr_behav1'] = (dat['usr_behav1']-mean)/sd
    

    return(dat)

In [24]:
# call function
def apply_function_on_table(table, name):
    table1 = table.apply(func=calc_metrics, axis=1)
    
    # change columns names
    names = [name + x for x in table1.columns.values]
    table1.columns = names
    
    return table1

In [13]:
# make calculations (~14 min calculation time)
print('PI')
pi_per_ssoid_met = apply_function_on_table( table = pi_per_ssoid_cont, name = 'PI_' )
print('V')
v_per_ssoid_met = apply_function_on_table( table = v_per_ssoid_cont, name = 'V_' )

print('PIPlus')
piPlus_per_ssoid_met = apply_function_on_table( table = piPlus_per_ssoid_cont, name = 'PIPlus_')
print('VPlus')
vPlus_per_ssoid_met = apply_function_on_table( table = vPlus_per_ssoid_cont, name = 'VPlus_')

print('PI/V')
pi_per_v_cont_met = apply_function_on_table( table = pi_per_v_cont, name = 'PI/V_')
print('PIPlus/V')
piPlus_per_v_cont_met = apply_function_on_table( table = piPlus_per_v_cont, name = 'PIPlus/V_')
print('VPlus/V')
vPlus_per_v_cont_met = apply_function_on_table( table = vPlus_per_v_cont, name = 'VPlus/V_')

PI


NameError: name 'apply_function_on_table' is not defined

In [34]:
# all tables will be joined
all_v = pi_per_ssoid_met.join(v_per_ssoid_met).join(piPlus_per_ssoid_met).join(vPlus_per_ssoid_met).join(pi_per_v_cont_met).join(piPlus_per_v_cont_met).join(vPlus_per_v_cont_met)

contract_dats =  last_contract_KTag[['WT SSOID', 'VE Jahr/KW', 'Kuendigungsstatus', 'Kuendigungsgrund',
       'Promotion ID', 'Vertragsform', 'Trial Kennzeichen',
       'Vertragsphase', 'VB Tag', 'VB Jahr', 'VB KW', 'VB Jahr/KW',
       'VE Tag', 'VE Jahr', 'VE KW',
       'K Tag']]
contract_dats.index = contract_dats['WT SSOID']

all_v_cont = contract_dats.join(all_v)

## Save Data

In [35]:
import pickle
all_v_cont.to_pickle('visits_contract')

In [12]:
import pickle
last_contract_KTag.to_pickle('contract_data')